### DISCUSSION TITLE
Running and debugging MOOSE from vs code or visual studio

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#18600] [Q&A Getting Started] [answered]
Hi

Can I run and debug MOOSE using vs code and/or visual studio? I knew that people use vs code for running and debugging on Linux through windows whether the project is stored on the local machine or on online. I was hoping the same for MOOSE but I am having some trouble. Is there a specific step-by-step procedure to do that? did anyone try to use vs code or visual studio to run ad debug moose application? 

Another issue, I tried to use the GDB on one example, I followed the instruction in this [page](https://mooseframework.inl.gov/application_development/debugging.html), whenever I tried to put a breakpoint at MPI_Abort, it gives this message, 

**(gdb) b MPI_Abort
Function "MPI_Abort" not defined.
Make breakpoint pending on future shared library load? (y or [n])** 

Thanks in advance.

**Comment by YaqiWang** [comment]
Reply only the second part of your questions: did you try run the example with something like `mpirun -np 2 ...`? I am asking to make sure you indeed have MPI library linked in.

**Reply by aaelmeli** [reply]
Yes, it can run on multiple cores when I used mpirun -np ....., however, when I use ./run_mytest -j nprocs -i -mtinput_file.i, it just runs over one core. I think this is the problem, but I do not know how to fix it. So, if you please can give a pointer, it will be great.

Abdo.

**Comment by GiudGiud** [comment, marked as answer]
@permcody @friedmud know of any users on VS?

**Reply by cticenhour** [reply]
@dylanjm and @dschwen were working on using MOOSE / MOOSE apps with visual studio at one point in FY20 I think.

**Reply by doquang** [reply]
> @dylanjm and @dschwen were working on using MOOSE / MOOSE apps with visual studio at one point in FY20 I think.

Can we compile Moose with visual studio as a native code, or we use visual studio for Moose on WSL?

**Reply by dschwen** [reply]
> Can we compile Moose with visual studio as a native code, or we use visual studio for Moose on WSL?

Use WSL. We rely heavily on Posix compatibility. Native Windows has a very different API.

**Reply by GiudGiud** [reply]
@fdkong any of your ORNL colleagues working with VS?
@dschwen ping again for the OP's question specifically

**Reply by fdkong** [reply]
No, he does not do any development at all. He is kind of an end-user.

**Comment by permcody** [comment]
Another good solution for Windows users is deploying MOOSE applications in a Docker container. This is something we may look at in the near future.

**Comment by friedmud** [comment]
I think the only viable path, for now, is using WSL... and that means _not_ using VS.  We have had several people report that WSL works.

**Comment by Hejie-Guo** [comment]
"(gdb) b MPI_Abort
Function "MPI_Abort" not defined.
Make breakpoint pending on future shared library load? (y or [n])"

Can anyone answer the question?

**Reply by GiudGiud** [reply]
press y for yes and it will break as expected
